cmd/traffic/cmd/manager/internal/ambassadoragent/cloudtoken/service.go

```diff
@@ -0,0 +1,121 @@
+package cloudtoken
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/datawire/dlib/dlog"
+	"github.com/datawire/k8sapi/pkg/k8sapi"
+	"github.com/telepresenceio/telepresence/v2/cmd/traffic/cmd/manager/managerutil"
+	apiv1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type Service interface {
+	MaybeAddToken(ctx context.Context, apikey string) error
+}
+
+type createConfigmapIfNotPresent struct {
+	createToken func(ctx context.Context, apikey string) error
+	done        <-chan struct{}
+}
+
+var (
+	suffix               = "agent-cloud-token"
+	cloudConnectTokenKey = "CLOUD_CONNECT_TOKEN"
+)
+
+// NewCreateConfigmapIfNotPresent will start up watchers to watch for a configmap or secret
+// with the apikey. If one is found, the watchers shut down. If MaybeAddToken is called
+// and successfully creates a configmap, the watchers are shut down.
+func NewCreateConfigmapIfNotPresent(ctx context.Context) *createConfigmapIfNotPresent {
+	managerns := managerutil.GetEnv(ctx).ManagerNamespace
+	clientset := k8sapi.GetK8sInterface(ctx)
+	watchers := newTokenWatchers(clientset, managerns)
+
+	// ctx for watchers, cancelled if token is found
+	cancelCtx, cancel := context.WithCancel(ctx)
+
+	go func() {
+		// search checks if apikey token exists, if it does, shut down the watchers
+		search := func() {
+			configmaps, err := watchers.mapsWatcher.List(cancelCtx)
```

Comment:
There are two watchers, and one cond (one subscribe), which means that when something changes, you need to search both resources. Wouldn't it be simpler and more efficient to have two separate watchers, subscribers, and listers?

Reply:
probably yes